src/maybe.h

```diff
@@ -134,8 +134,14 @@ class maybe_t : private maybe_detail::conditionally_copyable_t<T> {
     // return whether the receiver contains a value.
     bool has_value() const { return impl_.filled; }
 
-    // bool conversion indicates whether the receiver contains a value.
-    explicit operator bool() const { return impl_.filled; }
+    // A bool operator as a shortcut to test if the maybe_t has a value.
+    // Not enabled if the type T is already bool-convertible to prevent accidental misuse,
+    // otherwise the "typename std::enable_if<....>::type" evaluates to bool, giving us a definition
+    // of `explicit operator bool() const { ... }`
+    template <typename U = T> explicit operator
+    typename std::enable_if<!std::is_convertible<U, bool>::value, bool>::type() const {
+        return impl_.filled;
```

Comment:
Technically we only had problems with `maybe_t<bool>`, not with `maybe_t<size_t>` etc.
I guess `maybe_t<SomeType*>` is dangerous too but I think that happens to us only for typedef'd pointers (though I didn't check).

Reply:
We've actually had problems with `maybe_t<int>` (or maybe it was `size_t`), where an attempt to test for a non-zero offset was actually testing `maybe_t::has_value()`.